indexer/mailbox/src/message.rs

```diff
@@ -0,0 +1,136 @@
+    /// 4   Destination domain ID
+    pub destination: u32,
+    /// 32  Address in destination convention
+    pub recipient: H256,
+    /// 0+  Message contents
+    pub body: Vec<u8>,
+}
+
+impl From<RawHyperlaneMessage> for HyperlaneMessage {
+    fn from(m: RawHyperlaneMessage) -> Self {
+        HyperlaneMessage::from(&m)
+    }
+}
+
+impl From<&RawHyperlaneMessage> for HyperlaneMessage {
+    fn from(m: &RawHyperlaneMessage) -> Self {
+        let version = m[0];
+        let nonce: [u8; 4] = m[1..5].try_into().unwrap();
+        let origin: [u8; 4] = m[5..9].try_into().unwrap();
+        let sender: [u8; 32] = m[9..41].try_into().unwrap();
+        let destination: [u8; 4] = m[41..45].try_into().unwrap();
+        let recipient: [u8; 32] = m[45..77].try_into().unwrap();
+        let body = m[77..].try_into().unwrap();
+        Self {
+            version,
+            nonce: u32::from_be_bytes(nonce),
+            origin: u32::from_be_bytes(origin),
+            sender: H256::from(sender),
+            destination: u32::from_be_bytes(destination),
+            recipient: H256::from(recipient),
+            body,
+        }
+    }
+}
+
+impl Encode for HyperlaneMessage {
+    fn write_to<W>(&self, writer: &mut W) -> std::io::Result<usize>
+    where
+        W: std::io::Write,
+    {
+        writer.write_all(&self.version.to_be_bytes())?;
+        writer.write_all(&self.nonce.to_be_bytes())?;
+        writer.write_all(&self.origin.to_be_bytes())?;
+        writer.write_all(self.sender.as_ref())?;
+        writer.write_all(&self.destination.to_be_bytes())?;
+        writer.write_all(self.recipient.as_ref())?;
+        writer.write_all(&self.body)?;
+        Ok(HYPERLANE_MESSAGE_PREFIX_LEN + self.body.len())
+    }
+}
+
+impl Decode for HyperlaneMessage {
+    fn read_from<R>(reader: &mut R) -> Result<Self, DecodeError>
+    where
+        R: std::io::Read,
+    {
+        let mut version = [0u8; 1];
+        reader.read_exact(&mut version)?;
+
+        let mut nonce = [0u8; 4];
+        reader.read_exact(&mut nonce)?;
+
+        let mut origin = [0u8; 4];
+        reader.read_exact(&mut origin)?;
+
+        let mut sender = H256::zero();
+        reader.read_exact(sender.as_mut())?;
+
+        let mut destination = [0u8; 4];
+        reader.read_exact(&mut destination)?;
+
+        let mut recipient = H256::zero();
+        reader.read_exact(recipient.as_mut())?;
+
+        let mut body = vec![];
+        reader.read_to_end(&mut body)?;
+
+        Ok(Self {
+            version: u8::from_be_bytes(version),
+            nonce: u32::from_be_bytes(nonce),
+            origin: u32::from_be_bytes(origin),
+            sender,
+            destination: u32::from_be_bytes(destination),
+            recipient,
+            body,
+        })
+    }
+}
+
+impl HyperlaneMessage {
+    /// Convert the message to a message id
+    pub fn id(&self) -> H256 {
+        H256::from_slice(Keccak256::new().chain(self.to_vec()).finalize().as_slice())
```

Comment:
bit hard to read

Reply:
Also copied directly from the monorepo so I'm inclined to leave as is

But what I can do instead is do what I mention here https://github.com/hyperlane-xyz/fuel-contracts/pull/39/files/ec0ceaccbbbd620dbb78207bb1d2735c98a0f287#diff-c1df921b29777e810d4612bf6e315c71cf5f404d83cf0413cdeea3b4d2134aceR10 and move these into their own crate in the monorepo so that we don't need to copy and paste these